pipelines/controller/src/main/java/com/google/fhir/analytics/PipelineManager.java

```diff
@@ -382,12 +405,20 @@ synchronized void runIncrementalPipeline()
     mergerOptions.setDwh2(incrementalDwhRoot);
     mergerOptions.setMergedDwh(finalDwhRoot);
     mergerOptions.setRunner(FlinkRunner.class);
-    mergerOptions.setNumShards(dataProperties.getMaxWorkers());
-    FlinkPipelineOptions flinkOptions = mergerOptions.as(FlinkPipelineOptions.class);
-    flinkOptions.setFasterCopy(true);
-    flinkOptions.setMaxParallelism(dataProperties.getMaxWorkers());
+    // The number of shards is set based on the parallelism available for the FlinkRunner Pipeline
+    int numShards =
+        dataProperties.getNumThreads() > 0
+            ? dataProperties.getNumThreads()
+            : Runtime.getRuntime().availableProcessors();
+    mergerOptions.setNumShards(numShards);
+    FlinkPipelineOptions flinkOptionsForMerge = mergerOptions.as(FlinkPipelineOptions.class);
+    if (!Strings.isNullOrEmpty(flinkConfiguration.getFlinkConfDir())) {
+      flinkOptionsForMerge.setFlinkConfDir(flinkConfiguration.getFlinkConfDir());
+    }
+    flinkOptionsForMerge.setFasterCopy(true);
+    flinkOptionsForMerge.setMaxParallelism(dataProperties.getMaxWorkers());
     if (dataProperties.getNumThreads() > 0) {
-      flinkOptions.setParallelism(dataProperties.getNumThreads());
+      flinkOptionsForMerge.setParallelism(dataProperties.getNumThreads());
```

Comment:
Yes, this could be set to `numShards`, but I felt leaving it to Flink would be best. Internally Flink uses the similar logic to determine parallelism if the numThreads is not > 0. Setting this to `numShards` may be consistent so that the number of Shards and parallelism is always the same (Its the same even now as well).

Reply:
I see, so even when `numShards` is equal to the number of pipeline threads in ParquetIO, we still get 4 reshuffles, correct?

Reply:
Yes, right. The number of reshuffle operations for writing to new parquet files remains the same (i.e 4), irrespective of the `numShards`.